File: payment-core/go-services/internal/security/token_vault.go
package security

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"strings"
	"sync"
	"time"
)

type TokenDataType string

const (
	TokenTypePAN           TokenDataType = "pan"
	TokenTypeCVV           TokenDataType = "cvv"
	TokenTypePIN           TokenDataType = "pin"
	TokenTypeAccountNumber TokenDataType = "account_number"
	TokenTypeBVN           TokenDataType = "bvn"
	TokenTypeNIN           TokenDataType = "nin"
	TokenTypePII           TokenDataType = "pii"
	TokenTypeBiometric     TokenDataType = "biometric"
)

type TokenizedData struct {
	Token     string            `json:"token"`
	DataType  TokenDataType     `json:"data_type"`
	CreatedAt time.Time         `json:"created_at"`
	ExpiresAt *time.Time        `json:"expires_at,omitempty"`
	Metadata  map[string]string `json:"metadata,omitempty"`
}

type storedToken struct {
	EncryptedData string
	DataType      TokenDataType
	Metadata      map[string]string
	CreatedAt     time.Time
	ExpiresAt     *time.Time
}

type VaultKeyMetadata struct {
	KeyID     string     `json:"key_id"`
	Version   int        `json:"version"`
	Algorithm string     `json:"algorithm"`
	CreatedAt time.Time  `json:"created_at"`
	RotatedAt *time.Time `json:"rotated_at,omitempty"`
	Status    string     `json:"status"`
}

type keyData struct {
	Key      []byte
	Metadata VaultKeyMetadata
}

type AuditEntry struct {
	Timestamp time.Time     `json:"timestamp"`
	Operation string        `json:"operation"`
	TokenID   string        `json:"token_id,omitempty"`
	KeyID     string        `json:"key_id,omitempty"`
	DataType  TokenDataType `json:"data_type,omitempty"`
	Success   bool          `json:"success"`
	Error     string        `json:"error,omitempty"`
	Purpose   string        `json:"purpose,omitempty"`
}

type TokenVault struct {
	mu               sync.RWMutex
	tokenStore       map[string]*storedToken
	keyStore         map[string]*keyData
	auditLog         []AuditEntry
	currentDataKeyID string
	masterKeyID      string
}

func NewTokenVault() (*TokenVault, error) {
	tv := &TokenVault{
		tokenStore:       make(map[string]*storedToken),
		keyStore:         make(map[string]*keyData),
		auditLog:         make([]AuditEntry, 0),
		masterKeyID:      "master-key-v1",
		currentDataKeyID: "data-key-v1",
	}

	if err := tv.initializeKeys(); err != nil {
		return nil, fmt.Errorf("failed to initialize keys: %w", err)
	}

	return tv, nil
}

func (tv *TokenVault) initializeKeys() error {
	masterKey := make([]byte, 32)
	if _, err := rand.Read(masterKey); err != nil {
		return err
	}
	tv.keyStore[tv.masterKeyID] = &keyData{
		Key: masterKey,
		Metadata: VaultKeyMetadata{
			KeyID:     tv.masterKeyID,
			Version:   1,
			Algorithm: "AES-256-GCM",
			CreatedAt: time.Now(),
			Status:    "active",
		},
	}

	dataKey := make([]byte, 32)
	if _, err := rand.Read(dataKey); err != nil {
		return err
	}
	tv.keyStore[tv.currentDataKeyID] = &keyData{
		Key: dataKey,
		Metadata: VaultKeyMetadata{
			KeyID:     tv.currentDataKeyID,
			Version:   1,
			Algorithm: "AES-256-GCM",
			CreatedAt: time.Now(),
			Status:    "active",
		},
	}

	return nil
}

func (tv *TokenVault) getDataKey() ([]byte, error) {
	tv.mu.RLock()
	defer tv.mu.RUnlock()

	kd, ok := tv.keyStore[tv.currentDataKeyID]
	if !ok {
		return nil, errors.New("data encryption key not found")
	}
	return kd.Key, nil
}

func (tv *TokenVault) encrypt(plaintext string) (string, error) {
	key, err := tv.getDataKey()
	if err != nil {
		return "", err
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return "", err
	}

	ciphertext := gcm.Seal(nonce, nonce, []byte(plaintext), nil)
	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

func (tv *TokenVault) decrypt(encryptedData string) (string, error) {
	key, err := tv.getDataKey()
	if err != nil {
		return "", err
	}

	ciphertext, err := base64.StdEncoding.DecodeString(encryptedData)
	if err != nil {
		return "", err
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	nonceSize := gcm.NonceSize()
	if len(ciphertext) < nonceSize {
		return "", errors.New("ciphertext too short")
	}

	nonce, ciphertext := ciphertext[:nonceSize], ciphertext[nonceSize:]
	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return "", err
	}

	return string(plaintext), nil
}

func (tv *TokenVault) generateToken(dataType TokenDataType) string {
	prefix := tv.getTokenPrefix(dataType)
	randomBytes := make([]byte, 16)
	rand.Read(randomBytes)
	randomPart := hex.EncodeToString(randomBytes)

	hash := sha256.Sum256([]byte(randomPart))
	checksum := hex.EncodeToString(hash[:])[:4]

	return fmt.Sprintf("%s_%s%s", prefix, randomPart, checksum)
}

func (tv *TokenVault) getTokenPrefix(dataType TokenDataType) string {
	prefixes := map[TokenDataType]string{
		TokenTypePAN:           "tok_pan",
		TokenTypeCVV:           "tok_cvv",
		TokenTypePIN:           "tok_pin",
		TokenTypeAccountNumber: "tok_acc",
		TokenTypeBVN:           "tok_bvn",
		TokenTypeNIN:           "tok_nin",
		TokenTypePII:           "tok_pii",
		TokenTypeBiometric:     "tok_bio",
	}
	if p, ok := prefixes[dataType]; ok {
		return p
	}
	return "tok_unk"
}

type TokenizeOptions struct {
	ExpiresInSeconds int
	Metadata         map[string]string
}

func (tv *TokenVault) Tokenize(sensitiveData string, dataType TokenDataType, opts *TokenizeOptions) (*TokenizedData, error) {
	tv.mu.Lock()
	defer tv.mu.Unlock()

	token := tv.generateToken(dataType)
	encryptedData, err := tv.encrypt(sensitiveData)
	if err != nil {
		tv.logTokenOperation("tokenize", token, dataType, false, err.Error(), "")
		return nil, fmt.Errorf("encryption failed: %w", err)
	}

	stored := &storedToken{
		EncryptedData: encryptedData,
		DataType:      dataType,
		CreatedAt:     time.Now(),
	}

	if opts != nil {
		stored.Metadata = opts.Metadata
		if opts.ExpiresInSeconds > 0 {
			expiresAt := time.Now().Add(time.Duration(opts.ExpiresInSeconds) * time.Second)
			stored.ExpiresAt = &expiresAt
		}
	}

	tv.tokenStore[token] = stored
	tv.logTokenOperation("tokenize", token, dataType, true, "", "")

	result := &TokenizedData{
		Token:     token,
		DataType:  dataType,
		CreatedAt: stored.CreatedAt,
		ExpiresAt: stored.ExpiresAt,
	}
	if opts != nil {
		result.Metadata = opts.Metadata
	}

	return result, nil
}

type DetokenizeResult struct {
	Success bool
	Data    string
	Error   string
}

func (tv *TokenVault) Detokenize(token, purpose string) *DetokenizeResult {
	tv.mu.Lock()
	defer tv.mu.Unlock()

	stored, ok := tv.tokenStore[token]
	if !ok {
		tv.logTokenOperation("detokenize", token, "", false, "token not found", purpose)
		return &DetokenizeResult{Success: false, Error: "token not found"}
	}

	if stored.ExpiresAt != nil && stored.ExpiresAt.Before(time.Now()) {
		delete(tv.tokenStore, token)
		tv.logTokenOperation("detokenize", token, stored.DataType, false, "token expired", purpose)
		return &DetokenizeResult{Success: false, Error: "token expired"}
	}

	decrypted, err := tv.decrypt(stored.EncryptedData)
	if err != nil {
		tv.logTokenOperation("detokenize", token, stored.DataType, false, err.Error(), purpose)
		return &DetokenizeResult{Success: false, Error: "decryption failed"}
	}

	tv.logTokenOperation("detokenize", token, stored.DataType, true, "", purpose)
	return &DetokenizeResult{Success: true, Data: decrypted}
}

func (tv *TokenVault) DeleteToken(token string) bool {
	tv.mu.Lock()
	defer tv.mu.Unlock()

	_, existed := tv.tokenStore[token]
	delete(tv.tokenStore, token)
	tv.logTokenOperation("delete", token, "", existed, "", "")
	return existed
}

func (tv *TokenVault) RotateDataKey() (*VaultKeyMetadata, error) {
	tv.mu.Lock()
	defer tv.mu.Unlock()

	oldKeyID := tv.currentDataKeyID
	oldKey, ok := tv.keyStore[oldKeyID]
	if ok {
		oldKey.Metadata.Status = "pending_rotation"
	}

	newVersion := 1
	if oldKey != nil {
		newVersion = oldKey.Metadata.Version + 1
	}

	newKeyID := fmt.Sprintf("data-key-v%d", newVersion)
	newKey := make([]byte, 32)
	if _, err := rand.Read(newKey); err != nil {
		return nil, err
	}

	tv.keyStore[newKeyID] = &keyData{
		Key: newKey,
		Metadata: VaultKeyMetadata{
			KeyID:     newKeyID,
			Version:   newVersion,
			Algorithm: "AES-256-GCM",
			CreatedAt: time.Now(),
			Status:    "active",
		},
	}

	tv.currentDataKeyID = newKeyID

	if oldKey != nil {
		oldKey.Metadata.Status = "retired"
		now := time.Now()
		oldKey.Metadata.RotatedAt = &now
	}

	tv.logKeyOperation("rotate", newKeyID, newVersion)
	return &tv.keyStore[newKeyID].Metadata, nil
}

func (tv *TokenVault) GetKeyMetadata(keyID string) *VaultKeyMetadata {
	tv.mu.RLock()
	defer tv.mu.RUnlock()

	if keyID == "" {
		keyID = tv.currentDataKeyID
	}

	if kd, ok := tv.keyStore[keyID]; ok {
		return &kd.Metadata
	}
	return nil
}

func (tv *TokenVault) ListKeys() []VaultKeyMetadata {
	tv.mu.RLock()
	defer tv.mu.RUnlock()

	keys := make([]VaultKeyMetadata, 0, len(tv.keyStore))
	for _, kd := range tv.keyStore {
		keys = append(keys, kd.Metadata)
	}
	return keys
}

func (tv *TokenVault) logTokenOperation(operation, token string, dataType TokenDataType, success bool, errMsg, purpose string) {
	truncatedToken := token
	if len(token) > 20 {
		truncatedToken = token[:20] + "..."
	}

	entry := AuditEntry{
		Timestamp: time.Now(),
		Operation: operation,
		TokenID:   truncatedToken,
		DataType:  dataType,
		Success:   success,
		Error:     errMsg,
		Purpose:   purpose,
	}

	tv.auditLog = append(tv.auditLog, entry)

	if len(tv.auditLog) > 10000 {
		tv.auditLog = tv.auditLog[len(tv.auditLog)-10000:]
	}
}

func (tv *TokenVault) logKeyOperation(operation, keyID string, version int) {
	entry := AuditEntry{
		Timestamp: time.Now(),
		Operation: fmt.Sprintf("key_%s", operation),
		KeyID:     keyID,
		Success:   true,
	}
	tv.auditLog = append(tv.auditLog, entry)
}

func (tv *TokenVault) GetAuditLog(limit int) []AuditEntry {
	tv.mu.RLock()
	defer tv.mu.RUnlock()

	if limit <= 0 || limit > len(tv.auditLog) {
		limit = len(tv.auditLog)
	}

	start := len(tv.auditLog) - limit
	if start < 0 {
		start = 0
	}

	result := make([]AuditEntry, limit)
	copy(result, tv.auditLog[start:])
	return result
}

func MaskPAN(pan string) string {
	if len(pan) < 13 {
		return "****"
	}
	return pan[:6] + strings.Repeat("*", len(pan)-10) + pan[len(pan)-4:]
}

func MaskAccountNumber(accountNumber string) string {
	if len(accountNumber) < 6 {
		return "****"
	}
	return strings.Repeat("*", len(accountNumber)-4) + accountNumber[len(accountNumber)-4:]
}

func MaskBVN(bvn string) string {
	if len(bvn) != 11 {
		return "***********"
	}
	return bvn[:3] + "*****" + bvn[8:]
}

func GenerateHMAC(data string, key []byte) string {
	h := hmac.New(sha256.New, key)
	h.Write([]byte(data))
	return hex.EncodeToString(h.Sum(nil))
}

func VerifyHMAC(data, signature string, key []byte) bool {
	expected := GenerateHMAC(data, key)
	return hmac.Equal([]byte(expected), []byte(signature))
}

File: payment-switch/server/security/tokenVault.ts
import crypto from 'crypto';

export interface TokenizedData {
  token: string;
  dataType: TokenDataType;
  createdAt: Date;
  expiresAt?: Date;
  metadata?: Record<string, string>;
}

export type TokenDataType = 'pan' | 'cvv' | 'pin' | 'account_number' | 'bvn' | 'nin' | 'pii' | 'biometric';

export interface DetokenizeResult {
  success: boolean;
  data?: string;
  error?: string;
}

export interface HSMConfig {
  provider: 'aws_kms' | 'azure_keyvault' | 'gcp_kms' | 'hashicorp_vault' | 'local';
  keyId: string;
  region?: string;
  endpoint?: string;
}

export interface KeyMetadata {
  keyId: string;
  version: number;
  algorithm: string;
  createdAt: Date;
  rotatedAt?: Date;
  status: 'active' | 'pending_rotation' | 'retired';
}

const tokenStore = new Map<string, { encryptedData: string; dataType: TokenDataType; metadata?: Record<string, string>; createdAt: Date; expiresAt?: Date }>();
const keyStore = new Map<string, { key: Buffer; metadata: KeyMetadata }>();

let masterKeyId = 'master-key-v1';
let currentDataKeyId = 'data-key-v1';

function initializeKeys(): void {
  if (!keyStore.has(masterKeyId)) {
    const masterKey = crypto.randomBytes(32);
    keyStore.set(masterKeyId, {
      key: masterKey,
      metadata: {
        keyId: masterKeyId,
        version: 1,
        algorithm: 'AES-256-GCM',
        createdAt: new Date(),
        status: 'active'
      }
    });
  }

  if (!keyStore.has(currentDataKeyId)) {
    const dataKey = crypto.randomBytes(32);
    keyStore.set(currentDataKeyId, {
      key: dataKey,
      metadata: {
        keyId: currentDataKeyId,
        version: 1,
        algorithm: 'AES-256-GCM',
        createdAt: new Date(),
        status: 'active'
      }
    });
  }
}

initializeKeys();

function getDataKey(): Buffer {
  const keyData = keyStore.get(currentDataKeyId);
  if (!keyData) {
    throw new Error('Data encryption key not found');
  }
  return keyData.key;
}

function encrypt(plaintext: string): { ciphertext: string; iv: string; authTag: string } {
  const key = getDataKey();
  const iv = crypto.randomBytes(16);
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
  
  let encrypted = cipher.update(plaintext, 'utf8', 'hex');
  encrypted += cipher.final('hex');
  const authTag = cipher.getAuthTag();

  return {
    ciphertext: encrypted,
    iv: iv.toString('hex'),
    authTag: authTag.toString('hex')
  };
}

function decrypt(ciphertext: string, iv: string, authTag: string): string {
  const key = getDataKey();
  const decipher = crypto.createDecipheriv('aes-256-gcm', key, Buffer.from(iv, 'hex'));
  decipher.setAuthTag(Buffer.from(authTag, 'hex'));
  
  let decrypted = decipher.update(ciphertext, 'hex', 'utf8');
  decrypted += decipher.final('utf8');
  
  return decrypted;
}

function generateToken(dataType: TokenDataType): string {
  const prefix = getTokenPrefix(dataType);
  const randomPart = crypto.randomBytes(16).toString('hex');
  const checksum = crypto.createHash('sha256').update(randomPart).digest('hex').slice(0, 4);
  return `${prefix}_${randomPart}${checksum}`;
}

function getTokenPrefix(dataType: TokenDataType): string {
  const prefixes: Record<TokenDataType, string> = {
    pan: 'tok_pan',
    cvv: 'tok_cvv',
    pin: 'tok_pin',
    account_number: 'tok_acc',
    bvn: 'tok_bvn',
    nin: 'tok_nin',
    pii: 'tok_pii',
    biometric: 'tok_bio'
  };
  return prefixes[dataType];
}

export function tokenize(
  sensitiveData: string,
  dataType: TokenDataType,
  options?: { expiresInSeconds?: number; metadata?: Record<string, string> }
): TokenizedData {
  const token = generateToken(dataType);
  const { ciphertext, iv, authTag } = encrypt(sensitiveData);
  const encryptedData = `${iv}:${authTag}:${ciphertext}`;

  const expiresAt = options?.expiresInSeconds 
    ? new Date(Date.now() + options.expiresInSeconds * 1000)
    : undefined;

  tokenStore.set(token, {
    encryptedData,
    dataType,
    metadata: options?.metadata,
    createdAt: new Date(),
    expiresAt
  });

  logTokenOperation('tokenize', token, dataType, true);

  return {
    token,
    dataType,
    createdAt: new Date(),
    expiresAt,
    metadata: options?.metadata
  };
}

export function detokenize(token: string, purpose: string): DetokenizeResult {
  const stored = tokenStore.get(token);
  
  if (!stored) {
    logTokenOperation('detokenize', token, 'unknown' as TokenDataType, false, 'Token not found');
    return { success: false, error: 'Token not found' };
  }

  if (stored.expiresAt && stored.expiresAt < new Date()) {
    tokenStore.delete(token);
    logTokenOperation('detokenize', token, stored.dataType, false, 'Token expired');
    return { success: false, error: 'Token expired' };
  }

  try {
    const [iv, authTag, ciphertext] = stored.encryptedData.split(':');
    const decrypted = decrypt(ciphertext, iv, authTag);
    
    logTokenOperation('detokenize', token, stored.dataType, true, undefined, purpose);
    
    return { success: true, data: decrypted };
  } catch (error) {
    logTokenOperation('detokenize', token, stored.dataType, false, (error as Error).message);
    return { success: false, error: 'Decryption failed' };
  }
}

export function deleteToken(token: string): boolean {
  const existed = tokenStore.has(token);
  tokenStore.delete(token);
  logTokenOperation('delete', token, 'unknown' as TokenDataType, existed);
  return existed;
}

export function getTokenMetadata(token: string): Omit<TokenizedData, 'token'> | null {
  const stored = tokenStore.get(token);
  if (!stored) return null;
  
  return {
    dataType: stored.dataType,
    createdAt: stored.createdAt,
    expiresAt: stored.expiresAt,
    metadata: stored.metadata
  };
}

export function rotateDataKey(): KeyMetadata {
  const oldKeyId = currentDataKeyId;
  const oldKey = keyStore.get(oldKeyId);
  
  if (oldKey) {
    oldKey.metadata.status = 'pending_rotation';
  }

  const newVersion = oldKey ? oldKey.metadata.version + 1 : 1;
  const newKeyId = `data-key-v${newVersion}`;
  const newKey = crypto.randomBytes(32);
  
  keyStore.set(newKeyId, {
    key: newKey,
    metadata: {
      keyId: newKeyId,
      version: newVersion,
      algorithm: 'AES-256-GCM',
      createdAt: new Date(),
      status: 'active'
    }
  });

  currentDataKeyId = newKeyId;

  if (oldKey) {
    oldKey.metadata.status = 'retired';
    oldKey.metadata.rotatedAt = new Date();
  }

  logKeyOperation('rotate', newKeyId, newVersion);

  return keyStore.get(newKeyId)!.metadata;
}

export function getKeyMetadata(keyId?: string): KeyMetadata | null {
  const id = keyId || currentDataKeyId;
  const keyData = keyStore.get(id);
  return keyData?.metadata || null;
}

export function listKeys(): KeyMetadata[] {
  return Array.from(keyStore.values()).map(k => k.metadata);
}

const auditLog: Array<{
  timestamp: Date;
  operation: string;
  tokenId?: string;
  keyId?: string;
  dataType?: TokenDataType;
  success: boolean;
  error?: string;
  purpose?: string;
  actor?: string;
}> = [];

function logTokenOperation(
  operation: string,
  token: string,
  dataType: TokenDataType,
  success: boolean,
  error?: string,
  purpose?: string
): void {
  auditLog.push({
    timestamp: new Date(),
    operation,
    tokenId: token.slice(0, 20) + '...',
    dataType,
    success,
    error,
    purpose
  });

  if (auditLog.length > 10000) {
    auditLog.splice(0, auditLog.length - 10000);
  }
}

function logKeyOperation(operation: string, keyId: string, version: number): void {
  auditLog.push({
    timestamp: new Date(),
    operation: `key_${operation}`,
    keyId,
    success: true
  });
}

export function getAuditLog(limit: number = 100): typeof auditLog {
  return auditLog.slice(-limit);
}

export function maskPAN(pan: string): string {
  if (pan.length < 13) return '****';
  return pan.slice(0, 6) + '*'.repeat(pan.length - 10) + pan.slice(-4);
}

export function maskAccountNumber(accountNumber: string): string {
  if (accountNumber.length < 6) return '****';
  return '*'.repeat(accountNumber.length - 4) + accountNumber.slice(-4);
}

export function maskBVN(bvn: string): string {
  if (bvn.length !== 11) return '***********';
  return bvn.slice(0, 3) + '*****' + bvn.slice(-3);
}

export class TokenVaultService {
  tokenize = tokenize;
  detokenize = detokenize;
  deleteToken = deleteToken;
  getTokenMetadata = getTokenMetadata;
  rotateDataKey = rotateDataKey;
  getKeyMetadata = getKeyMetadata;
  listKeys = listKeys;
  getAuditLog = getAuditLog;
  maskPAN = maskPAN;
  maskAccountNumber = maskAccountNumber;
  maskBVN = maskBVN;
}

let tokenVaultInstance: TokenVaultService | null = null;

export function getTokenVaultService(): TokenVaultService {
  if (!tokenVaultInstance) {
    tokenVaultInstance = new TokenVaultService();
  }
  return tokenVaultInstance;
}

export default TokenVaultService;
